File: lc49.lua
-- 方法1：排序法
function groupAnagrams(strs)
    local map = {}
    
    for _, str in ipairs(strs) do
        -- 将字符串转换为字符表并排序
        local chars = {}
        for i = 1, #str do
            chars[i] = str:sub(i, i)
        end
        table.sort(chars)
        
        -- 排序后的字符串作为 key
        local key = table.concat(chars)
        
        if not map[key] then
            map[key] = {}
        end
        table.insert(map[key], str)
    end
    
    -- 构建结果
    local result = {}
    for _, group in pairs(map) do
        table.insert(result, group)
    end
    
    return result
end

-- 方法2：计数法（性能更好）
function groupAnagrams2(strs)
    local map = {}
    
    for _, str in ipairs(strs) do
        -- 统计每个字符出现次数
        local count = {}
        for i = 1, 26 do
            count[i] = 0
        end
        
        for i = 1, #str do
            local c = str:byte(i) - string.byte('a') + 1
            count[c] = count[c] + 1
        end
        
        -- 将计数数组转换为字符串作为 key
        local key = table.concat(count, "#")
        
        if not map[key] then
            map[key] = {}
        end
        table.insert(map[key], str)
    end
    
    -- 构建结果
    local result = {}
    for _, group in pairs(map) do
        table.insert(result, group)
    end
    
    return result
end

-- 测试代码
local function printResult(result)
    print("[")
    for i, group in ipairs(result) do
        local items = {}
        for _, str in ipairs(group) do
            table.insert(items, '"' .. str .. '"')
        end
        print("  [" .. table.concat(items, ", ") .. "]")
    end
    print("]")
end

-- 测试用例
local test1 = {"eat", "tea", "tan", "ate", "nat", "bat"}
print("输入: {\"eat\", \"tea\", \"tan\", \"ate\", \"nat\", \"bat\"}")
print("\n方法1 - 排序法:")
printResult(groupAnagrams(test1))

print("\n方法2 - 计数法:")
printResult(groupAnagrams2(test1))

File: lc49.cpp
/*
给你一个字符串数组，请你将 字母异位词 组合在一起。可以按任意顺序返回结果列表。

==字母异位词== : 字母异位词是通过重新排列不同单词或短语的字母而形成的单词或短语，并使用所有原字母一次。

**示例 1:**

**输入:** strs = ["eat", "tea", "tan", "ate", "nat", "bat"]

**输出:** [["bat"],["nat","tan"],["ate","eat","tea"]]

**解释：**

- 在 strs 中没有字符串可以通过重新排列来形成 `"bat"`。
- 字符串 `"nat"` 和 `"tan"` 是字母异位词，因为它们可以重新排列以形成彼此。
- 字符串 `"ate"` ，`"eat"` 和 `"tea"` 是字母异位词，因为它们可以重新排列以形成彼此。

**示例 2:**

**输入:** strs = [""]

**输出:** [[""]]

**示例 3:**

**输入:** strs = ["a"]

**输出:** [["a"]]

**提示：**

- `1 <= strs.length <= 104`
- `0 <= strs[i].length <= 100`
- `strs[i]` 仅包含小写字母
*/

#include <iostream> 
#include <vector> 
#include <algorithm> 
#include <map> 
#include <unordered_map> 
using namespace std;

// map 15ms 
// unordered_map 7ms

class Solution {
    public:
        // 计数法
        vector<vector<string>> groupAnagrams(vector<string>& strs) {
            // 1. 构建索引
            unordered_map<string,vector<string>> tmp;
            for (const auto& str : strs)
            {
                string key(26,'0');
                for (const char& c : str )
                {
                    key[c-'a']++;
                }
                tmp[key].push_back(str);
            }

            // 2. 构建结果
            vector<vector<string>> result;
            result.reserve(tmp.size());
            for (auto& it : tmp)
            {
                result.push_back(std::move(it.second));
            }

            // 3. 返回结果
            return result;
        }

    // 排序法
    vector<vector<string>> groupAnagrams2(vector<string>& strs) {
        // 1. 构建索引
        unordered_map<string,vector<string>> tmp;
        for (const auto& str : strs)
        {
            string key = str;
            sort(key.begin(),key.end());
            tmp[key].push_back(str);
        }

        // 2. 构建结果
        vector<vector<string>> result;
        result.reserve(tmp.size());
        for (auto& it : tmp)
        {
            result.push_back(std::move(it.second));
        }

        // 3. 返回结果
        return result;
    }
};

int main() {
    Solution solution;
    vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
    vector<vector<string>> result = solution.groupAnagrams(strs);
    for (auto vec : result) {
        for (auto str : vec) {
            cout << str << " ";
        }
        cout << endl;
    }
    return 0;
}
